packages/core/src/RenderPipeline/BasicRenderPipeline.ts

```diff
@@ -138,87 +72,131 @@ export class BasicRenderPipeline {
-        if (clearFlags & CameraClearFlags.Color) {
-          if (background.mode === BackgroundMode.Sky) {
-            background.sky._render(context);
-          } else if (background.mode === BackgroundMode.Texture && background.texture) {
-            this._drawBackgroundTexture(engine, background);
-          }
-        }
-        cullingResults.transparentQueue.render(camera, pass.mask, PipelineStage.Forward);
+    const { engine, scene } = camera;
+    const { background } = scene;
+
+    const internalColorTarget = this._internalColorTarget;
+    const rhi = engine._hardwareRenderer;
+    const colorTarget = camera.renderTarget ?? internalColorTarget;
+    const colorViewport = internalColorTarget ? PipelineUtils.defaultViewport : camera.viewport;
+    const needFlipProjection = (camera.renderTarget && cubeFace == undefined) || internalColorTarget != undefined;
+
+    if (context.flipProjection !== needFlipProjection) {
+      context.applyVirtualCamera(camera._virtualCamera, needFlipProjection);
+      this._updateMVPShaderData(context);
+      // @todo: It is more appropriate to prevent duplication based on `virtualCamera` at `RenderQueue#render`.
+      engine._renderCount++;
+    }
+
+    rhi.activeRenderTarget(colorTarget, colorViewport, mipLevel, cubeFace);
+    const clearFlags = camera.clearFlags & ~(ignoreClear ?? CameraClearFlags.None);
+    const color = background.solidColor;
+    if (clearFlags !== CameraClearFlags.None) {
+      rhi.clearRenderTarget(camera.engine, clearFlags, color);
+    }
+
+    opaqueQueue.render(camera, PipelineStage.Forward);
+    alphaTestQueue.render(camera, PipelineStage.Forward);
+    if (clearFlags & CameraClearFlags.Color) {
+      if (background.mode === BackgroundMode.Sky) {
+        background.sky._render(context);
+      } else if (background.mode === BackgroundMode.Texture && background.texture) {
+        this._drawBackgroundTexture(engine, background);
       }
+    }
 
-      renderTarget?._blitRenderTarget();
-      renderTarget?.generateMipmaps();
+    // Copy opaque texture
+    if (camera.opaqueTextureEnabled) {
+      // Should blit to resolve the MSAA
+      colorTarget._blitRenderTarget();
+
+      const opaqueTexturePass = this._opaqueTexturePass;
+      opaqueTexturePass.onConfig(camera, colorTarget.getColorTexture(0));
+      opaqueTexturePass.onRender(context, cullingResults);
+
+      // Should revert to original render target
+      rhi.activeRenderTarget(colorTarget, colorViewport, mipLevel, cubeFace);
     }
 
-    pass.postRender(camera, cullingResults.opaqueQueue, cullingResults.alphaTestQueue, cullingResults.transparentQueue);
+    transparentQueue.render(camera, PipelineStage.Forward);
+
+    colorTarget?._blitRenderTarget();
+    colorTarget?.generateMipmaps();
+
+    if (internalColorTarget) {
+      PipelineUtils.blitTexture(engine, <Texture2D>internalColorTarget.getColorTexture(0), null, 0, camera.viewport);
```

Comment:
the blit target maybe `camera.renderTarget`,not always null

Reply:
If has camera.renderTarget, this func will not called